### ISSUE TITLE
New IL2055 warnings with latest linker

### ISSUE THREAD

**Issue by eerhardt** [mono/linker#1887]
Updating `dotnet/runtime` to the latest linker is failing to build because of the following errors:

```
C:\git\runtime2\src\libraries\Microsoft.CSharp\src\Microsoft\CSharp\RuntimeBinder\Semantics\Types\NullableType.cs(47,54): Trim analysis error IL2055: Microsoft.CSharp.RuntimeBinder.Semantics.NullableType.AssociatedSystemType.get: Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\BinaryExpression.cs(2237,17): Trim analysis error IL2055: System.Linq.Expressions.Expression.GetResultTypeOfShift(Type,Type): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Linq.Expressions\src\System\Dynamic\Utils\TypeUtils.cs(26,17): Trim analysis error IL2055: System.Dynamic.Utils.TypeUtils.GetNullableType(Type): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Private.DataContractSerialization\src\System\Runtime\Serialization\Json\JsonFormatWriterGenerator.cs(685,21): Trim analysis error IL2055: System.Runtime.Serialization.Json.JsonFormatWriterGenerator.CriticalHelper.UnwrapNullableObject(LocalBuilder): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Private.DataContractSerialization\src\System\Runtime\Serialization\Json\JsonFormatWriterGenerator.cs(683,21): Trim analysis error IL2055: System.Runtime.Serialization.Json.JsonFormatWriterGenerator.CriticalHelper.UnwrapNullableObject(LocalBuilder): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\Converters\Value\EnumConverterFactory.cs(21,13): Trim analysis error IL2055: System.Text.Json.Serialization.Converters.EnumConverterFactory.CreateConverter(Type,JsonSerializerOptions): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializerOptions.Converters.cs(99,20): Trim analysis error IL2055: System.Text.Json.JsonSerializerOptions.<GetDictionaryKeyConverter>g__GetEnumConverter|4_0(<>c__DisplayClass4_0&): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\JsonStringEnumConverter.cs(57,13): Trim analysis error IL2055: System.Text.Json.Serialization.JsonStringEnumConverter.CreateConverter(Type,JsonSerializerOptions): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
C:\git\runtime2\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\Converters\Value\NullableConverterFactory.cs(36,13): Trim analysis error IL2055: System.Text.Json.Serialization.Converters.NullableConverterFactory.CreateValueConverter(Type,JsonConverter): Call to 'System.Type.MakeGenericType(Type[])' can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type. [C:\git\runtime2\src\libraries\src.proj]
```

Looking at one of them - `C:\git\runtime2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\BinaryExpression.cs(2237,17)`:

```C#
        private static Type GetResultTypeOfShift(Type left, Type right)
        {
            if (!left.IsNullableType() && right.IsNullableType())
            {
                // lift the result type to Nullable<T>
                return typeof(Nullable<>).MakeGenericType(left);
            }
            return left;
        }
```

This is calling `typeof(Nullable<>).MakeGenericType`, which should be safe because `Nullable<>` doesn't have any trimming annotations on its generic type.

Looking at other warnings, it seems to be warning whenever `where T : struct`. There is also a warning on using this in MakeGenericType:

```C#
internal class NullableConverter<T> : JsonConverter<T?> where T : struct
```
The linker shouldn't be warning in these cases.

cc @vitek-karas

**Comment by vitek-karas** [comment]
#1880 added a missing check that generic parameters on the open generic type used with `MakeGenericType` must not have `new()` constraint, since linker won't be able to validate that the type passed in fulfills the requirement of having parameterless constructor. Basically having the `new()` constraint is equivalent to having `DynamicallyAccessedMembers(PublicParameterlessConstructor)`.

The `struct` constraint implies the `new()` constraint:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/where-generic-type-constraint
>  The `new()` constraint can't be combined with the `struct` or `unmanaged` constraints. All types satisfying those constraints must have an accessible parameterless constructor, making the `new()` constraint redundant.

So the above case with `Nullable<T>` - it's declared as `public struct Nullable<T> where T : struct` and thus it also has `new()` constraint on the `T`. Linker has no way to validate the `MakeGenericType` in this case.

**Comment by vitek-karas** [comment]
/cc @MichalStrehovsky 

The new proposed "Fix" in #1888 will take care of the problem for all cases of `typeof(Nullable<>).MakeGenericType)` - as a special case.

That should solve all the above warnings with the exception of the EnumConverter related ones. I honestly don't know why `EnumConverter` even has the `struct` constraint (since it also has `Enum` constraint). Regardless, these cases are safe since the class will only work on enums, which are always "newable". The only downside is that we will have to suppress these warnings, which is not ideal, but there's no good way to handle this (other than adding more special casing logic for Enum constraints, but that feels way too corner case to me).

**Comment by MichalStrehovsky** [comment]
> I honestly don't know why EnumConverter even has the struct constraint (since it also has Enum constraint)

The `enum` constraint means "System.Enum or deriving from System.Enum". The `struct, enum` constraint means "just enums".

The constraint is probably forced because there's a call to `Enum.TryParse<T>`.

The constraint on that one is annoying. We should just delete it (https://github.com/dotnet/runtime/issues/48310#issuecomment-779443371).
